src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToInvocations.Analyzer.cs

```diff
@@ -0,0 +1,267 @@
+        private static bool VerifyAncestorOnlyHasOneCancellationTokenAsLastArgument(
+            INamedTypeSymbol cancellationTokenType,
+            IMethodSymbol methodDeclaration,
+            [NotNullWhen(returnValue: true)] out string? cancellationTokenParameterName)
+        {
+            if (methodDeclaration.Parameters.Count(x => x.Type.Equals(cancellationTokenType)) == 1 &&
+                methodDeclaration.Parameters.Last() is IParameterSymbol lastParameter &&
+                lastParameter.Type.Equals(cancellationTokenType)) // Covers the case when using an alias for ct
+            {
+                cancellationTokenParameterName = lastParameter.Name;
+                return true;
+            }
+
+            cancellationTokenParameterName = null;
+            return false;
+        }
+
+        // Checks if the invocation has an optional ct argument at the end or a params ct array at the end.
+        private static bool InvocationMethodTakesAToken(IMethodSymbol method, ImmutableArray<IArgumentOperation> arguments, INamedTypeSymbol cancellationTokenType)
+        {
+            return
+                !method.Parameters.IsEmpty &&
+                method.Parameters[method.Parameters.Length - 1] is IParameterSymbol lastParameter &&
+                (InvocationIgnoresOptionalCancellationToken(lastParameter, arguments, cancellationTokenType) ||
+                InvocationIsUsingParamsCancellationToken(lastParameter, arguments, cancellationTokenType));
+        }
+
+        // Check if the currently used overload is the one that takes the ct, but is utilizing the default value offered in the method signature.
+        // We want to offer a diagnostic for this case, so the user explicitly passes the ancestor's ct.
+        private static bool InvocationIgnoresOptionalCancellationToken(IParameterSymbol lastParameter, ImmutableArray<IArgumentOperation> arguments, INamedTypeSymbol cancellationTokenType)
+        {
+            if (lastParameter.Type.Equals(cancellationTokenType) &&
+                lastParameter.IsOptional) // Has a default value being used
+            {
+                // Find out if the ct argument is using the default value
+                return arguments.Any(x =>
+                    x.Parameter.Type.Equals(cancellationTokenType) &&
+                    x.ArgumentKind == ArgumentKind.DefaultValue); // The default value is being used
+            }
+
+            return false;
+        }
+
+        // Checks if the method has a `params CancellationToken[]` argument in the last position and ensure no ct is being passed.
+        private static bool InvocationIsUsingParamsCancellationToken(IParameterSymbol lastParameter, ImmutableArray<IArgumentOperation> arguments, INamedTypeSymbol cancellationTokenType)
+        {
+            if (lastParameter.IsParams &&
+                   lastParameter.Type.Kind == SymbolKind.ArrayType &&
```

Comment:
Isn't this check redundant? Can we have a non-array type params argument?
```suggestion
```

Reply:
It is actually redundant given your next line `lastParameter.Type is IArrayTypeSymbol arrayTypeSymbol`

Reply:
Good point, removed.